### PR TITLE
[WIP] Fix stack overflows in main game loop

### LINKED ISSUE #250: Stack overflow
# Details

Original Elona source code has many many many `goto` statements. Foobar currently emulates them by calling functions. It causes stack overflow quickly. Thus, Foobar throws exception and catches it in 
 function in higher position of call graph. We understand it is a very dirty hack, but I had no choice before.

This implementation causes stack overflow now. It is time to remove all harmful `goto` statements.

### PR DESCRIPTION
**Pull request by Ruin0x11**
# Related Issues

Closes #250.

# Summary

Unrolls the recursive calls causing stack overflows in the main game logic, making the game far more playable.

## Changes

- Placed all turn-related logic into `turn_wrapper()`. Now functions that need to follow up some result with another control routine, like returning to `pc_turn()` after a user error, can return a `turn_result_t` specifying the next action to take.
- Added a struct `menu_result` to handle menu subroutines that can return either a `succeeded` flag indicating if an action in the menu succeeded, or a `turn_result_t`.
  + Functions using the stat shouldn't need to use the `turn_result_t`, and vice versa. Some number of `assert`s were added for the latter case.
  + Changed the return types of functions that need to execute another action to return `turn_result_t` or `menu_result`.
- Added a struct `talk_result_t` and `talk_wrapper()` for unrolling recursion in nested dialogue routines.
  + Changed return types of nested dialogue functions to return `talk_result_t`.
- Added `casino_wrapper()` for unrolling recursion in casino routines.
- Added `main_menu_wrapper()` and `main_menu_result_t` for unrolling recursion in the main menu.
  + Changed return types of main menu functions to return `main_menu_result_t`.
- Renamed many unknown label functions.
- Renamed a few functions to be consistent with return values/context (executable commands, menu commands that return success/failure, menu commands that return nothing).
- There is no longer a need to throw an `elona_turn_sequence` in order to unwind the stack. 
- Moved `main_loop()` to be called at the end of `initialize_game()`.
- Removed the need to call `std::exit(0)`.

## TODO
- [x] Remove recursion in `talk_npc()`.
- [x] Remove recursion in Blackjack-related routines.
- [x] Verify Blackjack still works as expected.
- [x] Remove recursion in main menu routines.
- [x] Verify that the main menu/character making process doesn't break on completion/canceling at various points/including pressing 'F1' for help where available.
- [x] Investigate two usages of `initialize_map()` outside the main game loop, both inside `initialize_economy()`.
  + Seems it only gets called on starting a new game, so it should be okay.
- [x] Verify that all logic is equivalent in the codebase to the previous version.
- [x] Verify dialogue special case in Pael's mother quest.
- [x] Potentially fix a special case in `feat_menu()` where pressing `F1` causes a flag to be set.
- [x] Make the naming scheme for menu/player command/internal logic functions consistent (`do_xxx_command`, `xxx_menu`, `do_xxx_internal`).
- [x] Remove extraneous `TODO` comments.
- [x] Give better names to some members of `turn_result_t`.

### CODE REVIEW

**Comment by Ruin0x11** [comment]
Question: Should the result of opening something from `autopick` be taken into account for the PC's turn? The player's turn can end early on an open command if the item is a shopkeeper's trunk or deck, for example.

File: elonacore.cpp [hunk 1]
```diff
@@ -54422,12 +54340,14 @@ void proc_autopick()
                 }
             }
             elona::ci = ci;
-            do_open_command();
+            // TODO should the result be used?
+            turn_result_t turn_result = do_open_command();
```
Comment (Ruin0x11) [inline elonacore.cpp]: Should the result of this command be taken into effect for the turn lifecycle? The player's turn can end early if the item being opened is a shopkeeper's trunk, deck or the fire giant's shackles, for example.

File: elonacore.cpp [hunk 2]
```diff
@@ -54415,7 +54332,7 @@ void proc_autopick()
                 }
             }
             elona::ci = ci;
-            do_open_command();
+            turn_result_t turn_result = do_open_command();
```
Comment (Ruin0x11) [inline elonacore.cpp]: Should the result of this command be taken into effect for the turn lifecycle? The player's turn can end early if the item being opened is a shopkeeper's trunk, deck or the fire giant's shackles, for example.

**Comment by Ruin0x11** [comment]
I playtested the major parts which might have failed, so I think this is good to merge.

File: init.cpp [hunk 3]
```diff
@@ -992,22 +978,96 @@ void main_title_menu()
...
+
+            // Finished initializing, start the game.
+
+        case main_menu_result_t::initialize_game:
+            result = main_menu_result_t::initialize_game;
+            finished = true;
+            break;
+        case main_menu_result_t::finish_elona:
+            result = main_menu_result_t::finish_elona;
+            finished = true;
```
Comment (ki-foobar) [inline init.cpp]: If you select `Exit` in main menu in debug build, app crashes by `assert(0)`. The cause is unintended fall through in this case clause.

Comment (Ruin0x11) [reply init.cpp]: Whoops, added.

**Comment by ki-foobar** [comment]
I merged. It may have some bugs since the amount of changes is very huge.
Please post new issues/pull requsts if problems are found. Thanks.

**Pull request merged**